packages/assets-controllers/src/MultichainBalancesController/MultichainBalancesController.ts

```diff
@@ -0,0 +1,390 @@
+
+    if (account.metadata.snap) {
+      const scope = getScopeForAccount(account);
+      const assetTypes = NETWORK_ASSETS_MAP[scope];
+
+      partialState.balances[account.id] = await this.#getBalances(
+        account.id,
+        account.metadata.snap.id,
+        assetTypes,
+      );
+    }
+
+    this.update((state: Draft<MultichainBalancesControllerState>) => {
+      state.balances = {
+        ...state.balances,
+        ...partialState.balances,
+      };
+    });
+  }
+
+  /**
+   * Updates the balances of one account. This method doesn't return
+   * anything, but it updates the state of the controller.
+   *
+   * @param accountId - The account ID.
+   */
+  async updateBalance(accountId: string) {
+    // NOTE: No need to track the account here, since we start tracking those when
+    // the "AccountsController:accountAdded" is fired.
+    await this.#tracker.updateBalance(accountId);
+  }
+
+  /**
+   * Updates the balances of all supported accounts. This method doesn't return
+   * anything, but it updates the state of the controller.
+   */
+  async updateBalances() {
+    await this.#tracker.updateBalances();
+  }
+
+  /**
+   * Checks for non-EVM accounts.
+   *
+   * @param account - The new account to be checked.
+   * @returns True if the account is a non-EVM account, false otherwise.
+   */
+  #isNonEvmAccount(account: InternalAccount): boolean {
+    return (
+      !isEvmAccountType(account.type) &&
+      // Non-EVM accounts are backed by a Snap for now
+      account.metadata.snap !== undefined
+    );
+  }
+
+  /**
+   * Handles changes when a new account has been added.
+   *
+   * @param account - The new account being added.
+   */
+  async #handleOnAccountAdded(account: InternalAccount) {
+    if (!this.#isNonEvmAccount(account)) {
+      // Nothing to do here for EVM accounts
+      return;
+    }
+
+    this.#tracker.track(account.id, this.#getBlockTimeFor(account));
+    // NOTE: Unfortunately, we cannot update the balance right away here, because
+    // messenger's events are running synchronously and fetching the balance is
+    // asynchronous.
+    // Updating the balance here would resume at some point but the event emitter
+    // will not `await` this (so we have no real control "when" the balance will
+    // really be updated), see:
+    // - https://github.com/MetaMask/core/blob/v213.0.0/packages/accounts-controller/src/AccountsController.ts#L1036-L1039
+  }
+
+  /**
+   * Handles changes when a new account has been removed.
+   *
+   * @param accountId - The account ID being removed.
+   */
+  async #handleOnAccountRemoved(accountId: string) {
+    if (this.#tracker.isTracked(accountId)) {
+      this.#tracker.untrack(accountId);
+    }
+
+    if (accountId in this.state.balances) {
+      this.update((state: Draft<MultichainBalancesControllerState>) => {
+        delete state.balances[accountId];
+        return state;
```

Comment:
Nit: This return statement can be removed
```suggestion
```

Reply:
done in [bcaee6c](https://github.com/MetaMask/core/pull/4965/commits/bcaee6c42b4c196809fe2c59cd611ca155794b28)